pkg/controller/seed-controller-manager/mla/ratelimit_cortex_controller.go

```diff
@@ -0,0 +1,248 @@
+	MaxSamplesPerQuery *int32 `yaml:"max_samples_per_query,omitempty"`
+	IngestionBurstSize *int32 `yaml:"ingestion_burst_size,omitempty"`
+	MaxSeriesTotal     *int32 `yaml:"max_series_per_user,omitempty"`
+}
+
+type overrides struct {
+	Overrides map[string]tenantOverride `yaml:"overrides"`
+}
+
+// ratelimitCortexReconciler stores necessary components that are required to manage MLA(Monitoring, Logging, and Alerting) setup.
+type ratelimitCortexReconciler struct {
+	ctrlruntimeclient.Client
+
+	log                       *zap.SugaredLogger
+	workerName                string
+	recorder                  record.EventRecorder
+	versions                  kubermatic.Versions
+	ratelimitCortexController *ratelimitCortexController
+}
+
+// Add creates a new MLA controller that is responsible for
+// managing Monitoring, Logging and Alerting for user clusters.
+func newRatelimitCortexReconciler(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	versions kubermatic.Versions,
+	ratelimitCortexController *ratelimitCortexController,
+) error {
+	client := mgr.GetClient()
+
+	reconciler := &ratelimitCortexReconciler{
+		Client: client,
+
+		log:                       log,
+		workerName:                workerName,
+		recorder:                  mgr.GetEventRecorderFor(ControllerName),
+		versions:                  versions,
+		ratelimitCortexController: ratelimitCortexController,
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.MLAAdminSetting{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to watch MLAAdminSetting: %w", err)
+	}
+
+	return err
+}
+
+func (r *ratelimitCortexReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	mlaAdminSetting := &kubermaticv1.MLAAdminSetting{}
+	if err := r.Get(ctx, request.NamespacedName, mlaAdminSetting); err != nil {
+		return reconcile.Result{}, ctrlruntimeclient.IgnoreNotFound(err)
+	}
+
+	if !mlaAdminSetting.DeletionTimestamp.IsZero() {
+		if err := r.ratelimitCortexController.handleDeletion(ctx, log, mlaAdminSetting); err != nil {
+			return reconcile.Result{}, fmt.Errorf("handling deletion: %w", err)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.ratelimitCortexController.ensureLimits(ctx, mlaAdminSetting); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to ensure limits: %w", err)
+	}
+
+	if !kubernetes.HasFinalizer(mlaAdminSetting, mlaFinalizer) {
+		kubernetes.AddFinalizer(mlaAdminSetting, mlaFinalizer)
+		if err := r.Update(ctx, mlaAdminSetting); err != nil {
+			return reconcile.Result{}, fmt.Errorf("updating finalizers: %w", err)
+		}
+	}
```

Comment:
I think finalizer should be added before `ensureLimits`. otherwise, if this `Update` failed, `MLAAdminSetting` will not have finalizer, and will not trigger `handleDeletion` during deletion

Reply:
oh, good catch, thx

Reply:
done